source/MRMesh/MRMarchingCubes.cpp

```diff
@@ -328,49 +328,92 @@ TriangulationPlan{}
 
 const std::array<OutEdge, size_t( NeighborDir::Count )> cPlusOutEdges { OutEdge::PlusX, OutEdge::PlusY, OutEdge::PlusZ };
 
-}
+class VolumeMesher
+{
+public:
+    template<typename V, typename Positioner>
+    static Expected<TriMesh> run( const V& volume, const MarchingCubesParams& params, Positioner&& positioner );
+
+private:
+    explicit VolumeMesher( const Vector3i & dims, const MarchingCubesParams& params );
+
+    template<typename V, typename Positioner>
+    Expected<void> firstPass_( const V& volume, Positioner&& positioner );
+
+    Expected<TriMesh> secondPass_();
+
+private:
+    VolumeIndexer indexer_;
+    const MarchingCubesParams& params_;
+    int blockCount_ = 0;
+    int layerPerBlockCount_ = 0;
+
+    std::vector<BitSet> invalids_; ///< invalid voxels in each layer
+    std::vector<BitSet> lowerIso_; ///< voxels with the values lower then params.iso
+
+    SeparationPointStorage sepStorage_;
+};
 
 template<typename V, typename Positioner>
-Expected<TriMesh> volumeToMesh( const V& volume, const MarchingCubesParams& params, Positioner&& positioner )
+Expected<TriMesh> VolumeMesher::run( const V& volume, const MarchingCubesParams& params, Positioner&& positioner )
 {
-    TriMesh result;
     if ( volume.dims.x <= 0 || volume.dims.y <= 0 || volume.dims.z <= 0 )
-        return result;
-
+        return TriMesh{};
     MR_TIMER
 
-    auto cachingMode = params.cachingMode;
-    if ( cachingMode == MarchingCubesParams::CachingMode::Automatic )
-    {
-        if constexpr ( std::is_same_v<V, FunctionVolume> || std::is_same_v<V, FunctionVolumeU8> )
-            cachingMode = MarchingCubesParams::CachingMode::Normal;
-        else
-            cachingMode = MarchingCubesParams::CachingMode::None;
-    }
+    VolumeMesher mesher( volume.dims, params );
+    if ( auto e = mesher.firstPass_( volume, std::forward<Positioner>( positioner ) ); !e )
+        return unexpected( std::move( e.error() ) );
+
+    // free input volume, since it will not be used below any more
+    if ( params.freeVolume )
+        params.freeVolume();
 
-    VolumeIndexer indexer( volume.dims );
+    return mesher.secondPass_();
```

Comment:
You might rewrite this block in this way:
```c++
return
    mesher.firstPass_( volume, std::forward<Positioner>( positioner ) )
    .and_then( [&]
    {
        // free input volume, since it will not be used below any more
        if ( params.freeVolume )
            params.freeVolume();
            
        return mesher.secondPass_();
    } );